### PR TITLE
Memoized SectionInfo component

### LINKED ISSUE #360: Improve SectionSelect performance
As discovered in #357, when courses have a lot of sections (i.e. MATH-151) it can cause the selecting of the sections perform pretty slowly

Gif of the problem. Note that I'm spamming click anytime I'm hovering over the sections

### PR DESCRIPTION
**Pull request by firejake308**
## Description

Spam-clicking a section in `SectionSelect` currently results in a really uncomfortable lag period where the UI is frozen. I originally attempted to fix this by virtualizing the list, but virtualization results ina an uncomfortable lag period where the `SectionSelect` is blank. I have found a better solution in memoization, which is basically the functional component equivalent of `shouldComponentUpdate`. Specifically, I moved the display for each section out of `SectionSelect` and into a new component called `SectionSelect`. Next, I memo-ized this `SectionInfo`, instructing it not to re-render unless it has been selected/de-selected. Significantly, this makes the assumption that none of the other fields or sub-fields in `SectionSelected` will ever change without also completely re-rendering the course card.

## Rationale

I had to make a new `ThunkAction` to toggle the selected state of each section due to some really weird behavior with JavaScript closures. If anyone thinks they understand closures, please explain how they work because apparently I don't.

## How to test

Spam-click a section. Should update quickly. Now try selecting multiple sections. Selecting a new section shouldn't un-select your previous choices.

## Screenshots

memo

## Related tasks

Closes #360

### CODE REVIEW

File: autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectCard/ExpandedCourseCard/SectionSelect/SectionSelect.tsx
Comment (ryan-conn) [inline autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectCard/ExpandedCourseCard/SectionSelect/SectionSelect.tsx:132]: I'm so glad this is a separate component now, no offense but this was an absolute mess before lol

File: autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectCard/ExpandedCourseCard/SectionSelect/SectionSelect.tsx
Comment (ryan-conn) [inline autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectCard/ExpandedCourseCard/SectionSelect/SectionSelect.tsx:15]: I thought the default equality function checked reference equality, why did you explicitly do it here?

File: autoscheduler/frontend/src/redux/actions/courseCards.ts
Comment (ryan-conn) [inline autoscheduler/frontend/src/redux/actions/courseCards.ts:242]: I'm fine with you moving this here since we don't need a selector in the `SectionInfo` component, and it abstracts out the logic a bit from the component itself

**Review by ryan-conn — CHANGES_REQUESTED** [review]
Couple of small things, great job making this a lot more responsive (and also making the sectionselect a lot more readable)! 

One thing I noticed is that it's still pretty slow when you switch between basic/section select since it has to rerender everything, it would be really nice if we could memoize it as well somehow (although I'm not sure we would do so right now). Definitely don't need to do that in this PR, but we should keep that in mind so that in the future we could potentially make it even faster.

Also I think we should wait for #384 to be merged before merging this so Shaham doesn't have to deal with rebasing his branch

File: autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectCard/ExpandedCourseCard/SectionSelect/SectionSelect.tsx
Comment (firejake308) [reply autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectCard/ExpandedCourseCard/SectionSelect/SectionSelect.tsx]: Nice catch. I think I put that in while debugging to see if it would change anything, and understandably, it didn't. Removed in latest commit.

**Review by gannonprudhomme — APPROVED** [review]
looks good to me!

**Pull request merged**
